internal/ratelimit/globallimiter.go

```
@@ -52,23 +56,30 @@ type globalRateLimiter struct {
 	prefix     string
 	bucketName string
 	pool       *redis.Pool
+	logger     log.Logger
 
 	// optionally used for tests
 	nowFunc func() time.Time
 	// optionally used for tests
 	timerFunc func(d time.Duration) (<-chan time.Time, func() bool)
 }
 
-func NewGlobalRateLimiter(bucketName string) GlobalLimiter {
+func NewGlobalRateLimiter(logger log.Logger, bucketName string) GlobalLimiter {
+	logger = logger.Scoped(fmt.Sprintf("GlobalRateLimiter.%s", bucketName), "")
 	pool, ok := kv().Pool()
 	if !ok {
-		return getInMemoryLimiter(bucketName)
+		rl := -1 // Documented default.
+		if rate := conf.Get().DefaultRateLimit; rate != nil {
+			rl = *rate
+		}
+		return getInMemoryLimiter(bucketName, rl)
```

Should we log here that the fallback is taken?

---

Also: should we really fallback when we can't construct the global one? What happens if Redis is temporarily unvailable and we fall back to the `InMemoryLimiter` here? Does it always stay in-memory then?

---

`!ok` is only truthy in App mode, where we want to fall back to the InMemoryLimiter - it doesn't mean that redis is unavailable temporarily, it means that no redis available ever.

---

Can't we make that more explicit then? And do `if envvar.IsApp()`? Also: the method name, `NewGlobalRateLimiter`, is confusing then if an expected outcome of calling the method is that it does NOT return a global rate limiter, but an in-memory one.

---

> Can't we make that more explicit then? And do if envvar.IsApp()? 

Hmm we could, this is the most accurate check we have today - if the deployment doesn't support redis we fall back to in-memory (see below why that is indeed a global limiter in app). We could use an app check and otherwise panic? I don't really want to return an error in that case, because it's a build issue.

> Also: the method name, NewGlobalRateLimiter, is confusing then if an expected outcome of calling the method is that it does NOT return a global rate limiter, but an in-memory one.

It is a global limiter in app mode, because app is a single binary this is shared across all services and also configured properly by the worker job. It just doesn't use redis.

---

Okay, that makes sense. But I still think `if isApp() { return in memory } else { return global-redis-limiter }` is clearest. Because app is the only deployment that we support that doesn't have Redis, right? Everything else should have Redis and if it doesn't, I think it be logged somewhere.

>  I don't really want to return an error in that case, because it's a build issue.

It's also a configuration issue, no? If someone configures the wrong redis_url in env vars? or someone turns redis off? in that case we should error, right?

---

Setting an invalid redis URL will not make `.Pool()` return false, 

```
const MemoryKeyValueURI = "redis+memory:memory"

const dbKeyValueURIScheme = "redis+postgres"
```

are the only two values that will make redis be in-memory or DB driven. 

I don't think anyone should use these except app and they are completely undocumented. Otherwise, we cannot implement anything close to this rate limiter.

---

we could do that if you find it clearer:

```diff
diff --git a/internal/ratelimit/globallimiter.go b/internal/ratelimit/globallimiter.go
index 7eb923539bf..960762cb912 100644
--- a/internal/ratelimit/globallimiter.go
+++ b/internal/ratelimit/globallimiter.go
@@ -17,6 +17,7 @@ import (
 	"golang.org/x/time/rate"
 
 	"github.com/sourcegraph/sourcegraph/internal/conf"
+	"github.com/sourcegraph/sourcegraph/internal/conf/deploy"
 	"github.com/sourcegraph/sourcegraph/internal/redispool"
 	"github.com/sourcegraph/sourcegraph/lib/errors"
 )
@@ -68,6 +69,10 @@ func NewGlobalRateLimiter(logger log.Logger, bucketName string) GlobalLimiter {
 	logger = logger.Scoped(fmt.Sprintf("GlobalRateLimiter.%s", bucketName), "")
 	pool, ok := kv().Pool()
 	if !ok {
+		if !deploy.IsApp() {
+			logger.Error("Redis pool not set, global rate limiter will not work")
+			return &notWorkingGlobalLimiter{}
+		}
 		rl := -1 // Documented default in site-config JSON schema. -1 means infinite.
 		if rate := conf.Get().DefaultRateLimit; rate != nil {
 			rl = *rate
```

---

Yeah, I think that's better, but again: I think making the assumption explicit is best. "If it's app, we use this, otherwise this". Because in your diff here, you essentially say `!ok == deploy.IsApp()`, so why not make that explicit vs. having two conditions?

---

Well, we can make the app check first and be very explicit about that, but we will still need a check for `!ok` and if it's not available, we would have to fail, making these config options (to use redis in memory or db) effectively forbidden options, because no rate limited activities would work anymore (ie. all code host connectivity). Not sure if that's great. 
I don't know why anyone should be allowed to use these, though.

---

```diff
diff --git a/internal/ratelimit/globallimiter.go b/internal/ratelimit/globallimiter.go
index 7eb923539bf..1e5136589c0 100644
--- a/internal/ratelimit/globallimiter.go
+++ b/internal/ratelimit/globallimiter.go
@@ -17,6 +17,7 @@ import (
 	"golang.org/x/time/rate"
 
 	"github.com/sourcegraph/sourcegraph/internal/conf"
+	"github.com/sourcegraph/sourcegraph/internal/conf/deploy"
 	"github.com/sourcegraph/sourcegraph/internal/redispool"
 	"github.com/sourcegraph/sourcegraph/lib/errors"
 )
@@ -66,8 +67,9 @@ type globalRateLimiter struct {
 
 func NewGlobalRateLimiter(logger log.Logger, bucketName string) GlobalLimiter {
 	logger = logger.Scoped(fmt.Sprintf("GlobalRateLimiter.%s", bucketName), "")
-	pool, ok := kv().Pool()
-	if !ok {
+
+	// App doesn't have a redis, so we use an in-memory implementation.
+	if deploy.IsApp() {
 		rl := -1 // Documented default in site-config JSON schema. -1 means infinite.
 		if rate := conf.Get().DefaultRateLimit; rate != nil {
 			rl = *rate
@@ -75,6 +77,12 @@ func NewGlobalRateLimiter(logger log.Logger, bucketName string) GlobalLimiter {
 		return getInMemoryLimiter(bucketName, rl)
 	}
 
+	pool, ok := kv().Pool()
+	if !ok {
+		logger.Error("Redis pool not set, global rate limiter will not work")
+		return &notWorkingGlobalLimiter{}
+	}
+
 	return &globalRateLimiter{
 		prefix:     tokenBucketGlobalPrefix,
 		bucketName: bucketName,
```

effectively, this

---

That I think is better

---

I don't get it: if Redis is configured to be in-memory or db-backend (never heard of that mode), isn't it still redis and couldn't we then still use it?

---

It's not a real redis, it's a mock implementation that was built for app which doesn't run a real redis server. It supports basic functionality like Get, Set and so forth, but does not support running lua scripts which we use for this rate limiter implementation. Thus, all implementations of `KeyValue` that don't support calling `.Pool()` are not usable for globallimiter.